### ISSUE TITLE
Make sure our code works without issues with WooCommerce's Custom Order Tables 

### ISSUE THREAD

**Issue by asumaran** [woocommerce/woocommerce-gateway-stripe#2407]
For more context visit: p1659937693943169-slack-C9QDV4TL4

As part of [WooCommerce's migration to Custom Order Tables](https://developer.woocommerce.com/2022/01/17/the-plan-for-the-woocommerce-custom-order-table/) (COT) we need to prepare our code to be compatible with it when it's released to everyone.

So far, while testing the [critical flows](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows) (No UPE) I've found an issue related to refunds.

<img src="https://user-images.githubusercontent.com/1025173/183561746-eb8d092b-85a9-42b4-b7eb-dcc08c220d70.png" width="400">

There are likely more errors than that. We must at least ensure the critical flows (with or without UPE enabled) runs as expected without issues.

Per the COT Recipe Book, [these are the functions](https://github.com/Automattic/woocommerce/wiki/COT-Upgrade-Recipe-Book#auditing-the-code-base-for-direct-db-access-usage) we need to replace with the new WooCommerce APIs for getting/setting post and post meta data.

- get_post
- get_post_field
- get_post_status
- get_post_type
- get_post_type_object
- get_posts
- metadata_exists
- get_post_meta
- get_metadata
- get_metadata_raw
- get_metadata_default
- get_metadata_by_mid
- wp_insert_post
- add_metadata
- add_post_meta
- wp_update_post
- update_post_meta
- update_metadata
- update_metadata_by_mid
- delete_metadata
- delete_post_meta
- delete_metadata_by_mid
- delete_post_meta_by_key
- wp_delete_post
- wp_trash_post
- wp_untrash_post
- wp_transition_post_status
- clean_post_cache
- update_post_caches
- update_postmeta_cache
- post_exists

We seem affected, given that there are multiple occurrences of some of these functions in our code base.

<details>
<summary>List of occurrences</summary>

```
Searching 249 files for "(get_post)|(get_post_field)|(get_post_status)|(get_post_type)|(get_post_type_object)|(get_posts)|(metadata_exists)|(get_post_meta)|(get_metadata)|(get_metadata_raw)|(get_metadata_default)|(get_metadata_by_mid)|(wp_insert_post)|(add_metadata)|(add_post_meta)|(wp_update_post)|(update_post_meta)|(update_metadata)|(update_metadata_by_mid)|(delete_metadata)|(delete_post_meta)|(delete_metadata_by_mid)|(delete_post_meta_by_key)|(wp_delete_post)|(wp_trash_post)|(wp_untrash_post)|(wp_transition_post_status)|(clean_post_cache)|(update_post_caches)|(update_postmeta_cache)|(post_exists)" (regex)

/Users/asumaran/www/woocommerce-gateway-stripe/includes/class-wc-gateway-stripe.php:
  861: 		clean_post_cache( $order->get_id() );

/Users/asumaran/www/woocommerce-gateway-stripe/includes/class-wc-stripe-helper.php:
  103: 		delete_post_meta( $order_id, self::META_NAME_FEE );
  104: 		delete_post_meta( $order_id, self::LEGACY_META_NAME_FEE );
  162: 		delete_post_meta( $order_id, self::META_NAME_NET );
  163: 		delete_post_meta( $order_id, self::LEGACY_META_NAME_NET );

/Users/asumaran/www/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php:
  409: 				'metadata'    => $gateway->get_metadata_from_order( $order ),

/Users/asumaran/www/woocommerce-gateway-stripe/includes/admin/class-wc-stripe-privacy.php:
  128: 							'value' => get_post_meta( $order->get_id(), '_stripe_source_id', true ),
  132: 							'value' => get_post_meta( $order->get_id(), '_stripe_customer_id', true ),
  193: 							'value' => get_post_meta( $subscription->get_id(), '_stripe_source_id', true ),
  197: 							'value' => get_post_meta( $subscription->get_id(), '_stripe_customer_id', true ),
  342: 		$stripe_source_id = get_post_meta( $subscription_id, '_stripe_source_id', true );
  361: 			delete_post_meta( $renewal_order_id, '_stripe_source_id' );
  362: 			delete_post_meta( $renewal_order_id, '_stripe_refund_id' );
  363: 			delete_post_meta( $renewal_order_id, '_stripe_customer_id' );
  366: 		delete_post_meta( $subscription_id, '_stripe_source_id' );
  367: 		delete_post_meta( $subscription_id, '_stripe_refund_id' );
  368: 		delete_post_meta( $subscription_id, '_stripe_customer_id' );
  381: 		$stripe_source_id   = get_post_meta( $order_id, '_stripe_source_id', true );
  382: 		$stripe_refund_id   = get_post_meta( $order_id, '_stripe_refund_id', true );
  383: 		$stripe_customer_id = get_post_meta( $order_id, '_stripe_customer_id', true );
  394: 		delete_post_meta( $order_id, '_stripe_source_id' );
  395: 		delete_post_meta( $order_id, '_stripe_refund_id' );
  396: 		delete_post_meta( $order_id, '_stripe_customer_id' );

/Users/asumaran/www/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php:
  384: 			update_post_meta( $subscription_id, '_stripe_customer_id', $source->customer );
  387: 				update_post_meta( $subscription_id, '_stripe_source_id', $source->payment_method );
  389: 				update_post_meta( $subscription_id, '_stripe_source_id', $source->source );
  400: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_customer_id' );
  401: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_source_id' );
  403: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_card_id' );
  405: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_intent_id' );
  419: 		delete_post_meta( $renewal_order->get_id(), '_stripe_intent_id' );
  433: 		update_post_meta( $subscription->get_id(), '_stripe_customer_id', $renewal_order->get_meta( '_stripe_customer_id', true ) );
  434: 		update_post_meta( $subscription->get_id(), '_stripe_source_id', $renewal_order->get_meta( '_stripe_source_id', true ) );
  449: 		$source_id       = get_post_meta( $subscription_id, '_stripe_source_id', true );
  453: 			$source_id = get_post_meta( $subscription_id, '_stripe_card_id', true );
  456: 			update_post_meta( $subscription_id, '_stripe_source_id', $source_id );
  457: 			delete_post_meta( $subscription_id, '_stripe_card_id', $source_id );
  463: 					'value' => get_post_meta( $subscription_id, '_stripe_customer_id', true ),
  530: 		$stripe_source_id = get_post_meta( $subscription->get_id(), '_stripe_source_id', true );
  534: 			$stripe_source_id = get_post_meta( $subscription->get_id(), '_stripe_card_id', true );
  537: 			update_post_meta( $subscription->get_id(), '_stripe_source_id', $stripe_source_id );
  541: 		$stripe_customer_id = get_post_meta( $subscription->get_id(), '_stripe_customer_id', true );
  560: 			$stripe_customer_id = get_post_meta( $subscription->get_parent_id(), '_stripe_customer_id', true );
  561: 			$stripe_source_id   = get_post_meta( $subscription->get_parent_id(), '_stripe_source_id', true );
  565: 				$stripe_source_id = get_post_meta( $subscription->get_parent_id(), '_stripe_card_id', true );
  568: 				update_post_meta( $subscription->get_parent_id(), '_stripe_source_id', $stripe_source_id );

/Users/asumaran/www/woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-multibanco.php:
  241: 		$data = get_post_meta( $order_id, '_stripe_multibanco', true );
  292: 		update_post_meta( $order_id, '_stripe_multibanco', $data );

/Users/asumaran/www/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php:
 1556: 		$posted_data     = $wc_checkout->get_posted_data();

/Users/asumaran/www/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
  572: 				$request['metadata'] = $this->get_metadata_from_order( $order );
 1306: 	public function get_metadata_from_order( $order ) {

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/class-wc-gateway-stripe.php:
  860: 		clean_post_cache( $order->get_id() );

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/class-wc-stripe-helper.php:
  103: 		delete_post_meta( $order_id, self::META_NAME_FEE );
  104: 		delete_post_meta( $order_id, self::LEGACY_META_NAME_FEE );
  162: 		delete_post_meta( $order_id, self::META_NAME_NET );
  163: 		delete_post_meta( $order_id, self::LEGACY_META_NAME_NET );

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php:
  409: 				'metadata'    => $gateway->get_metadata_from_order( $order ),

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/admin/class-wc-stripe-privacy.php:
  121: 							'value' => get_post_meta( $order->get_id(), '_stripe_source_id', true ),
  125: 							'value' => get_post_meta( $order->get_id(), '_stripe_customer_id', true ),
  186: 							'value' => get_post_meta( $subscription->get_id(), '_stripe_source_id', true ),
  190: 							'value' => get_post_meta( $subscription->get_id(), '_stripe_customer_id', true ),
  335: 		$stripe_source_id = get_post_meta( $subscription_id, '_stripe_source_id', true );
  354: 			delete_post_meta( $renewal_order_id, '_stripe_source_id' );
  355: 			delete_post_meta( $renewal_order_id, '_stripe_refund_id' );
  356: 			delete_post_meta( $renewal_order_id, '_stripe_customer_id' );
  359: 		delete_post_meta( $subscription_id, '_stripe_source_id' );
  360: 		delete_post_meta( $subscription_id, '_stripe_refund_id' );
  361: 		delete_post_meta( $subscription_id, '_stripe_customer_id' );
  374: 		$stripe_source_id   = get_post_meta( $order_id, '_stripe_source_id', true );
  375: 		$stripe_refund_id   = get_post_meta( $order_id, '_stripe_refund_id', true );
  376: 		$stripe_customer_id = get_post_meta( $order_id, '_stripe_customer_id', true );
  387: 		delete_post_meta( $order_id, '_stripe_source_id' );
  388: 		delete_post_meta( $order_id, '_stripe_refund_id' );
  389: 		delete_post_meta( $order_id, '_stripe_customer_id' );

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php:
  384: 			update_post_meta( $subscription_id, '_stripe_customer_id', $source->customer );
  387: 				update_post_meta( $subscription_id, '_stripe_source_id', $source->payment_method );
  389: 				update_post_meta( $subscription_id, '_stripe_source_id', $source->source );
  400: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_customer_id' );
  401: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_source_id' );
  403: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_card_id' );
  405: 		delete_post_meta( $resubscribe_order->get_id(), '_stripe_intent_id' );
  419: 		delete_post_meta( $renewal_order->get_id(), '_stripe_intent_id' );
  433: 		update_post_meta( $subscription->get_id(), '_stripe_customer_id', $renewal_order->get_meta( '_stripe_customer_id', true ) );
  434: 		update_post_meta( $subscription->get_id(), '_stripe_source_id', $renewal_order->get_meta( '_stripe_source_id', true ) );
  449: 		$source_id       = get_post_meta( $subscription_id, '_stripe_source_id', true );
  453: 			$source_id = get_post_meta( $subscription_id, '_stripe_card_id', true );
  456: 			update_post_meta( $subscription_id, '_stripe_source_id', $source_id );
  457: 			delete_post_meta( $subscription_id, '_stripe_card_id', $source_id );
  463: 					'value' => get_post_meta( $subscription_id, '_stripe_customer_id', true ),
  530: 		$stripe_source_id = get_post_meta( $subscription->get_id(), '_stripe_source_id', true );
  534: 			$stripe_source_id = get_post_meta( $subscription->get_id(), '_stripe_card_id', true );
  537: 			update_post_meta( $subscription->get_id(), '_stripe_source_id', $stripe_source_id );
  541: 		$stripe_customer_id = get_post_meta( $subscription->get_id(), '_stripe_customer_id', true );
  560: 			$stripe_customer_id = get_post_meta( $subscription->get_parent_id(), '_stripe_customer_id', true );
  561: 			$stripe_source_id   = get_post_meta( $subscription->get_parent_id(), '_stripe_source_id', true );
  565: 				$stripe_source_id = get_post_meta( $subscription->get_parent_id(), '_stripe_card_id', true );
  568: 				update_post_meta( $subscription->get_parent_id(), '_stripe_source_id', $stripe_source_id );

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-multibanco.php:
  237: 		$data = get_post_meta( $order_id, '_stripe_multibanco', true );
  288: 		update_post_meta( $order_id, '_stripe_multibanco', $data );

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php:
 1584: 		$posted_data     = $wc_checkout->get_posted_data();

/Users/asumaran/www/woocommerce-gateway-stripe/release/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
  534: 				$request['metadata'] = $this->get_metadata_from_order( $order );
 1268: 	public function get_metadata_from_order( $order ) {

/Users/asumaran/www/woocommerce-gateway-stripe/tests/phpunit/helpers/class-wc-helper-product.php:
  394: 		update_post_meta( $id, '_test2', 'world' );

103 matches across 17 files
```

</details>

**Comment by asumaran** [comment]
### Tests run with "Use the WooCommerce orders tables" option enabled

- [x] Merchant - Onboarding - Manual plugin installation and setup
- [x] Shopper - Checkout - Checkout with normal credit card
- [x] Shopper - Checkout - Save card at checkout
- [x] Shopper - Checkout - Checkout with saved card
- [x] Shopper - Checkout - Checkout with SCA card
- [x] Shopper - Checkout - Checkout failures (with various cards)
- [x] Shopper - Checkout - Checkout with Payment Request
- [ ] Shopper - Checkout - Checkout with Payment Request (3DS card)
  - Can't test this one.
- [x] Shopper - Payment methods - Add new payment method
- [x] Shopper - Payment methods - Delete payment method
- [x] Merchant - Order - Manual capture
- [ ] Merchant - Order - Full refund
  - Manual Refund: Error in Refunded, Net Payment, Paid, Stripe Payout amounts.
  - Via Stripe Refund: Error in Refunded, Net Payment, Paid, Stripe Payout amounts.
- [ ] Merchant - Order - Partial refunds (one, several)
  - Same errors as Full refund.
- [ ] Merchant - Order - Refund failure
  - Refund fails but refund amounts are incorrect
- [ ] Shopper - Subscriptions - Purchase subscription product
  - COT: "Error processing checkout. Please try again."
  - WordPress Posts Table: Works fine.
- [ ] Shopper - Subscriptions - Purchaser free trial subscription
  - COT: "Error processing checkout. Please try again." although the order is created.
  - WordPress Posts Table: Works fine.
- [ ] Shopper - Subscriptions - Purchase multiple subscriptions
  - COT: "Error processing checkout. Please try again." although the order is created.
  - WordPress Posts Table: Works fine.
- [x] Shopper - Subscriptions - Renew subscription
- [x] Shopper - Subscriptions - Change payment method to new card
- [x] Shopper - Subscriptions - Change payment method to saved card
- [x] Shopper - Subscriptions - Change default payment method
- [x] Shopper - Subscriptions - Set default payment method
- [x] Merchant - Subscriptions - Renew subscription manually
  - COT: Although creating the subscription didn't work the renewal worked fine.
- [x] Merchant - Subscriptions - Renew subscription automatically
  - COT: "Invalid Subscription."
- [x] Merchant - WooCommerce Blocks - Create a Cart page with the Cart Block
- [x] Merchant - WooCommerce Blocks - Create a Checkout page with the Checkout Block
- [x] Shopper - WooCommerce Blocks - Checkout with normal credit card

-- Starting from here "Keep the posts table and the orders tables synchronized" option was disabled --

- [x] Shopper - WooCommerce Blocks - Checkout with SCA card
- [x] Shopper - WooCommerce Blocks - Checkout failures (with various cards)
- [x] Shopper - WooCommerce Blocks - Checkout with Payment Request
- [x] Shopper - WooCommerce Blocks - Save card at checkout
- [x] Shopper - WooCommerce Blocks - Checkout with saved card
- [ ] Shopper - WooCommerce Blocks - Purchase subscription product
  - COT: Critical error
  - WordPress Posts Table: Works fine
- [ ] Shopper - WooCommerce Blocks - Purchase free trial subscription
  - COT: Critical error
  - WordPress Posts Table: Works fine
- [ ] Shopper - WooCommerce Blocks - Purchase multiple subscriptions
  - COT: Critical error
  - WordPress Posts Table: Works fine
- [ ] Shopper - APMs - Checkout with iDEAL
  - COT: Error
  - WordPress Posts Table: Works fine
- [x] Shopper - APMs - Checkout with Sofort
- [x] Shopper - APMs - Checkout with giropay
- [x] Shopper - APMs - Checkout with EPS
- [x] Shopper - APMs - Checkout with Bancontact
- [x] Shopper - APMs - Checkout with Przelewy24
- [x] Shopper - APMs - Checkout with SEPA
- [x] Shopper - APMs - Checkout with Multibanco
- [x] Shopper - APMs - Checkout with Alipay
- [x] Shopper - APMs - Checkout with Boleto
- [ ] Shopper - APMs - Checkout with OXXO
  - Wasn't able to test this one. Needs a mexican stripe account it seems.

**Comment by asumaran** [comment]
Refund related issues seem to be fixed now after @vedanshujain has merged `enhancement/33771` into the `cot-main` branch.

- [x] Merchant - Order - Full refund
- [x] Merchant - Order - Partial refunds (one, several)
- [x] Merchant - Order - Refund failure

**Comment by asumaran** [comment]
So I looked over the code and wanted to see which functions need to be updated or adjusted to work with COT. Here's the list:

- ~no ocurrences - get_post~
- ~no ocurrences - get_post_field~
- ~no ocurrences - get_post_status~
- ~no ocurrences - get_post_type~
- ~no ocurrences - get_post_type_object~
- ~no ocurrences - get_posts~
- ~no ocurrences - metadata_exists~
- **18 ocurrences - get_post_meta**
- ~no ocurrences - get_metadata~
- ~no ocurrences - get_metadata_raw~
- ~no ocurrences - get_metadata_default~
- ~no ocurrences - get_metadata_by_mid~
- ~no ocurrences - wp_insert_post~
- ~no ocurrences - add_metadata~
- ~no ocurrences - add_post_meta~
- ~no ocurrences - wp_update_post~
- **08 ocurrences - update_post_meta**
- ~no ocurrences - update_metadata~
- ~no ocurrences - update_metadata_by_mid~
- ~no ocurrences - delete_metadata~
- **19 ocurrences - delete_post_meta**
- ~no ocurrences - delete_metadata_by_mid~
- ~no ocurrences - delete_post_meta_by_key~
- ~no ocurrences - wp_delete_post~
- ~no ocurrences - wp_trash_post~
- ~no ocurrences - wp_untrash_post~
- ~no ocurrences - wp_transition_post_status~
- **01 ocurrences - clean_post_cache**
- ~no ocurrences - update_post_caches~
- ~no ocurrences - update_postmeta_cache~
- ~no ocurrences - post_exists~

So we need to take care of the `get_post_meta`, `update_post_meta`, `delete_post_meta`, and `clean_post_cache` functions only.

**Comment by asumaran** [comment]
I went through all the instances where get_post_meta, update_post_meta, delete_post_meta, and clean_post_cache are called and made sure it's working well with COT enabled and "Keep the posts table and the orders tables synchronized" unchecked.

Next step is to create a PR with tweaks for COT compatibility. Unfortunately, all the cases related to subscriptions are not supported for now because COT will not support WC Subscriptions. See pc2DNy-1V1-p2

# get_post_meta

## includes/admin/class-wc-stripe-privacy.php

- 2 instances in the `order_data_exporter` method.
  - How to test:
    - Export Personal Data > Add user email or billing email.
  - ✅ Result:
    Exported orders should contain `Stripe payment id` and `Stripe customer id` information
- 2 instances in the `subscriptions_data_exporter` method.
  - How to test:
    - Export Personal Data > Add user email or billing email.
  - ❌ Result:
    - COT is incompatible with WC Subscriptions (See pc2DNy-1V1-p2)
- 1 instance in the `maybe_handle_subscription` method.
  - How to test:
    - Erase Personal Data > Add user email or billing email.
  - ❌ Result:
    - COT is incompatible with WC Subscriptions
- 3 instances in the `maybe_handle_order` method.
  - How to test:
    - Erase Personal Data > Add user email or billing email.
  - ✅ Result:
    - Stripe data (_stripe_source_id, _stripe_refund_id and _stripe_customer_id) should be deleted from the order meta.

## includes/compat/trait-wc-stripe-subscriptions.php

- 3 instances in the `add_subscription_payment_meta` method.
  - How to test: 
    1. As a client buy a subscription
    2. As a merchant go to edit the subscription WooCommerce > Subscriptions > Click to edit the subscription.
    3. Click the pencil icon in "Billing"
    4. In "Payment method" "stripe" should be selected. "Stripe Customer ID" and "Stripe Source ID" should show correct information.
  - ❌ Result:
    - COT is incompatible with WC Subscriptions.
- 6 instances in the `maybe_render_subscription_payment_method` method.
  - How to test:
    1. Make sure auto renewal is enabled in the store.
    2. Buy a subscription
    3. Go to "My Account" > "My Subscription"
    4. Make sure "Auto renew" toggle is active
    5. Make sure a "Payment" method is displayed: "Via visa card ending in 4242".
    6. Notice that we may want to test the cases: "If we couldn't find a Stripe customer linked to the subscription, fallback to the user meta data." and "If we couldn't find a Stripe customer linked to the account, fallback to the order meta data."
  - ❌ Result:
    - COT is incompatible with WC Subscriptions.

## includes/payment-methods/class-wc-gateway-stripe-multibanco.php

[x] 1 instance in the `get_instructions` method.
  - How to test:
    1. Switch currency to Euro.
    2. Enable Multibanco.
    3. Checkout using Multibanco.
  - ✅ Results:
    - Confirm Multibanco info is displayed on the "Order received" page.
  - What was done:
    - Updated `get_instructions` to get the order meta using the new order API.

# update_post_meta

## includes/compat/trait-wc-stripe-subscriptions.php

- 3 instances in the `maybe_update_source_on_subscription_order` method.
  - How to test:
    1. Buy a subscription
    2. Make sure the checkout process finishes correctly
    3. Check the `_stripe_customer_id` and `_stripe_source_id` meta data are correctly added to the subscription in the DB (Use the subscription ID).
  - ❌ Result:
    - COT is incompatible with WC Subscriptions.
- 2 instances in the `update_failing_payment_method` method.
  - How to test:
    1. Not sure how to test this case. (it's using a subscription ID from a `WC_Subscription` object so it's probably compatible with COT).
  - ❌ Result:
    - COT is incompatible with WC Subscriptions.
- 1 instance in the `add_subscription_payment_meta` method.
  - How to test:
    1. As a client buy a subscription
    2. As a merchant go to edit the subscription WooCommerce > Subscriptions > Click to edit the subscription.
    3. Click the pencil icon in "Billing"
    4. In "Payment method" "stripe" should be selected. "Stripe Customer ID" and "Stripe Source ID" should show correct information.
    5. Notice the `update_post_meta` is there for the following cases:
      - "For BW compat will remove in future." which hapens when there's no `_stripe_source_id` attached to the subscription.
      - And "If we couldn't find a Stripe customer linked to the account, fallback to the order meta data." which happens when there's no `_stripe_source_id` attached to the account and there's no `_stripe_source_id` attached the parent order of the subscription.
  - ❌ Result:
    - COT is incompatible with WC Subscriptions.
- 2 instances in the `maybe_render_subscription_payment_method` method.
  - How to test:
    1. Make sure auto renewal is enabled in the store.
    2. Buy a subscription
    3. Go to "My Account" > "My Subscription"
    4. Make sure "Auto renew" toggle is active
    5. Make sure a "Payment" method is displayed: "Via visa card ending in 4242".
    6. Notice that we may want to test the cases: "If we couldn't find a Stripe customer linked to the subscription, fallback to the user meta data." and "If we couldn't find a Stripe customer linked to the account, fallback to the order meta data."
  - ❌ Result:
    - COT is incompatible with WC Subscriptions.

## includes/payment-methods/class-wc-gateway-stripe-multibanco.php

[x] 1 instance in the `save_instructions` method.
  - How to test:
    1. Switch currency to Euro.
    2. Enable Multibanco.
    3. Checkout using Multibanco.
  - ✅ Results:
    - Check `_stripe_multibanco` meta key is saved on the `wp_wc_orders_meta` with "order_id = {recently created order ID}"

## tests/phpunit/helpers/class-wc-helper-product.php

[x] 1 instance in the `save_post_test_update_meta_data_direct` method
  - How to test:
    1. Unit tests should run without issues.
  - Results: ✅ 
  - What was done: Removed the method because it was not used on WCStripe.

# delete_post_meta

## includes/class-wc-stripe-helper.php

- 2 instances in `delete_stripe_fee` method.
  - How to test:
    1. Make sure auto renewal is enabled in the store.
    2. Enable "Accept Early Renewal Payments via a Modal" in WooCommerce > Subscriptions
    3. Buy a subscription
    4. Go to My Account > My Subscription.
    5. Click "Renew now".
    6. Click "Pay now".
    7. This should trigger a call to the `delete_stripe_fee` the problem I see is that there's no `_stripe_fee` attached to the order. And when the renewal process finishes there's no `_stripe_fee` meta data attached to the order. So, not sure if this `delete_stripe_fee` is doing anything at all.
  - ❌ Result: COT is incompatible with WC Subscriptions.
- 2 instances in `delete_stripe_net` method.
  - How to test:
    1. Make sure auto renewal is enabled in the store.
    2. Enable "Accept Early Renewal Payments via a Modal" in WooCommerce > Subscriptions
    3. Buy a subscription
    4. Go to My Account > My Subscription.
    5. Click "Renew now".
    6. Click "Pay now".
    7. Same as the previous steps, there's no indication this method doing anything. There's no `_stripe_net` in the order then `_stripe_net` is added to the order.
  - ❌ Result: Doesn't work because COT is incompatible with WC Subscriptions.

## includes/admin/class-wc-stripe-privacy.php

- 6 instances in `maybe_handle_subscription` method.
  - How to test:
    - Add an early `return true` to the `is_retention_expired` in the `includes/admin/class-wc-stripe-privacy.php` file.
    - As a buyer:
      - Buy a subscription
      - Go to My Account > My Subscription
      - Renew the subscription
      - Cancel the subscription
    - As a merchant:
      - Go to Tools > Erase Personal Data.
      - Request a data erasure using the buyer's email.
      - Click "Force erase personal data".
      - Erasure should complete successfully.
      - `_stripe_source_id`, `_stripe_refund_id`, and `_stripe_customer_id` should be removed from the renewal order and the subscription.
  - ❌ Result: COT is incompatible with WC Subscriptions.
- 3 instances in the `maybe_handle_order` method.
  - How to test: Erase Personal Data > Add user email or billing email.
  - ✅ Result: Stripe data (_stripe_source_id, _stripe_refund_id and _stripe_customer_id) should be deleted from the order meta.

## includes/compat/trait-wc-stripe-subscriptions.php

- 4 instances in the `delete_resubscribe_meta`
  - How to test:
    - `delete_resubscribe_meta` is added as a callback of 
    - `wcs_resubscribe_order_created` which is called by
    - `wcs_create_resubscribe_order` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/wcs-resubscribe-functions.php` which is called by
    - `generate_renewal_order` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-renewal-order.php` which is called by
    - `generate_renewal_orders` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php` which is called by
    - `upgrade_really_old_versions` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php` which is called by
    - `ajax_upgrade` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php` which is called by
    - `init` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php`
    - Not sure how to trigger this to test `delete_resubscribe_meta` since it's called only when upgrading old subscritions.
  - ❌ Result: COT is incompatible with WC Subscriptions.
- 1 instance in the `delete_renewal_meta` method.
  - How to test:
    There's two paths that call `delete_renewal_meta`
    1. Path 1
      - `delete_renewal_meta` is called by
      - `delete_resubscribe_meta` in `includes/compat/trait-wc-stripe-subscriptions.php` which is added as callback of action
      - `wcs_resubscribe_order_created` in `includes/compat/trait-wc-stripe-subscriptions.php` which is called by
      - `wcs_create_resubscribe_order` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/wcs-resubscribe-functions.php` which is called by
      - `generate_renewal_order` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-renewal-order.php` which is called by
      - `generate_renewal_orders` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php` which is called by
      - `upgrade_really_old_versions` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php` which is called by
      - `ajax_upgrade` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php` which is called by
      - `init` in `woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/upgrades/class-wc-subscriptions-upgrader.php`
      - Not sure how to trigger this to test `delete_resubscribe_meta` since it's called only when upgrading old subscritions.
    2. Path 2
      - `delete_renewal_meta` is added as a callback of action
      - `wcs_renewal_order_created` which is called by
        - `wcs_create_renewal_order` called in multiple places by
          - first:
            - `process_early_renewal_request` is added as a callback of action
            - `wp_loaded` which is added in
            - `init` in `WCS_Early_Renewal_Modal_Handler` which is called by
            - `init` in `WC_Subscriptions_Plugin` which is called by
            - `init` in `WC_Subscriptions` which is called by
            - `woocommerce-subscriptions.php` file when loaded.
          - second:
            - `process_renewal` in `/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-manager.php` which is called by
            - `prepare_renewal`  in `/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-manager.php` which is added as callback of action
            - `woocommerce_scheduled_subscription_payment` in `/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-manager.php` which is added by
            - `init` in `WC_Subscriptions_Manager` in `/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-manager.php` which is called by
            - `init` in `WC_Subscriptions_Core_Plugin` in `/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-core-plugin.php` which is called by
            - `init` in `WC_Subscriptions_Plugin` in `/woocommerce-subscriptions/includes/class-wc-subscriptions-plugin.php` which is called by
            - `init` in `WC_Subscriptions` in `/woocommerce-subscriptions/woocommerce-subscriptions.php` which is called by
            - `woocommerce-subscriptions.php` file when loaded.
          - there's a bunch of other places but testing one should be enough.
  - ❌ Result: COT is incompatible with WC Subscriptions.
- 1 instance in the `add_subscription_payment_meta` method.
  - How to test:
    - `add_subscription_payment_meta` is added as a callback of action
    - `woocommerce_subscription_payment_meta` which is added in 
    - `maybe_init_subscriptions` of trait `WC_Stripe_Subscriptions_Trait` in `/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php`
    - action is called in:
     - `update_payment_method`
     - `create_subscription`
     - `register_routes` via API
     - seems to be called when creating subscriptions via the REST API
  - ❌ Result: COT is incompatible with WC Subscriptions.

# clean_post_cache

## includes/class-wc-gateway-stripe.php

-  1 instance in the `verify_intent_after_checkout` method.
  - How to test:
    1. Log in as a buyer
    1. Make an order
    2. Go to My Account > Order and click any order to see its details.
    3. The `woocommerce_account_view-order_endpoint` should be called which calls `verify_intent_after_checkout` which then calls the `clean_post_cache` function.
  - Result:
  - ✅ Result: Seems to be working fine. Although we might want to clean the cache of the order more properly.

**Comment by cesarcosta99** [comment]
Doing another battery of tests with “Use the WooCommerce orders tables” option enabled:

**Store setup:**

1. WordPress 6.1.1
2. WooCommerce 7.2.2
3. WooCommerce Blocks 9.3.0
4. WooCommerce Stripe `trunk` (7f3cbdd9)
5. WooCommerce Subscriptions 4.7.0

- [x] Merchant - Onboarding - Manual plugin installation and setup
- [x] Shopper - Checkout - Checkout with normal credit card
- [x] Shopper - Checkout - Save card at checkout
- [x] Shopper - Checkout - Checkout with saved card
- [x] Shopper - Checkout - Checkout with SCA card
- [x] Shopper - Checkout - Checkout failures (with various cards)
- [x] Shopper - Checkout - Checkout with Payment Request
- [ ] Shopper - Checkout - Checkout with Payment Request (3DS card)
  - Can’t test this one.
- [x] Shopper - Payment methods - Add new payment method
- [x] Shopper - Payment methods - Delete payment method
- [x] Merchant - Order - Manual capture
- [x] Merchant - Order - Full refund
- [x] Merchant - Order - Partial refund
- [x] Merchant - Order - Refund failure
- [x] Shopper - Subscriptions - Purchase subscription product
- [x] Shopper - Subscriptions - Purchase free trial subscription
- [x] Shopper - Subscriptions - Purchase multiple subscriptions (same schedule)
- [x] Shopper - Subscriptions - Purchase multiple subscriptions (differing schedules)
- [x] Shopper - Subscriptions - Renew subscription
- [ ] Shopper - Subscriptions - Change payment method to new card
  - Payment method is not updated after landing on Subscription view, even though alert says so.
- [ ] Shopper - Subscriptions - Change payment method to saved card
  - Same as previous.
- [ ] Shopper - Subscriptions - Set default payment method
  - Same as previous.
- [ ] Shopper - Subscriptions - Change default payment method
  - Unable to test because of previous outcome.
- [x] Merchant - Subscriptions - Renew subscription automatically
- [x] Merchant - Subscriptions - Renew subscription manually
- [x] Merchant - WooCommerce Blocks - Create a Cart page with the Cart Block
- [x] Merchant - WooCommerce Blocks - Create a Checkout page with the Checkout Block
- [x] Shopper - WooCommerce Blocks - Checkout with normal credit card
- [x] Shopper - WooCommerce Blocks - Checkout with SCA card
- [x] Shopper - WooCommerce Blocks - Checkout failures (with various cards)
- [x] Shopper - WooCommerce Blocks - Checkout with Payment Request
- [x] Shopper - WooCommerce Blocks - Save card at checkout
- [x] Shopper - WooCommerce Blocks - Checkout with saved card
- [x] Shopper - WooCommerce Blocks - Purchase subscription product
- [x] Shopper - WooCommerce Blocks - Purchase free trial subscription
- [x] Shopper - WooCommerce Blocks - Purchase multiple subscriptions (same schedule)
- [x] Shopper - WooCommerce Blocks - Purchase multiple subscriptions (differing schedules)
- [x] Shopper - APMs - Checkout with iDEAL
- [x] Shopper - APMs - Checkout with Sofort
- [x] Shopper - APMs - Checkout with giropay
- [x] Shopper - APMs - Checkout with EPS
- [x] Shopper - APMs - Checkout with Bancontact
- [x] Shopper - APMs - Checkout with Przelewy24
- [x] Shopper - APMs - Checkout with SEPA
- [x] Shopper - APMs - Checkout with Multibanco
- [x] Shopper - APMs - Checkout with Alipay
- [x] Shopper - APMs - Checkout with Boleto
- [x] Shopper - APMs - Checkout with OXXO

---

Starting from here “Keep the posts table and the orders tables synchronized” option was disabled.

- [x] Shopper - WooCommerce Blocks - Checkout with SCA card
- [x] Shopper - WooCommerce Blocks - Checkout failures (with various cards)
- [x] Shopper - WooCommerce Blocks - Checkout with Payment Request
- [x] Shopper - WooCommerce Blocks - Save card at checkout
- [x] Shopper - WooCommerce Blocks - Checkout with saved card
- [x] Shopper - WooCommerce Blocks - Purchase subscription product
- [x] Shopper - WooCommerce Blocks - Purchase free trial subscription
- [x] Shopper - WooCommerce Blocks - Purchase multiple subscriptions (same schedule)
- [x] Shopper - WooCommerce Blocks - Purchase multiple subscriptions (differing schedules)
- [x] Shopper - APMs - Checkout with iDEAL
- [x] Shopper - APMs - Checkout with Sofort
- [x] Shopper - APMs - Checkout with giropay
- [x] Shopper - APMs - Checkout with EPS
- [x] Shopper - APMs - Checkout with Bancontact
- [x] Shopper - APMs - Checkout with Przelewy24
- [x] Shopper - APMs - Checkout with SEPA
- [x] Shopper - APMs - Checkout with Multibanco
- [x] Shopper - APMs - Checkout with Alipay
- [x] Shopper - APMs - Checkout with Boleto
- [x] Shopper - APMs - Checkout with OXXO

**Result:**

No issues related to COT were found. When “Keep the posts table and the orders tables synchronized” was enabled,  both `wp_posts` and `wp_wc_orders` have stayed in sync; after disabling it, only `wp_wc_orders` was having information being set – as expected.

**Bugs found:**

- ~Payment methods are being saved without consent. Sometimes the save checkbox doesn’t display, then the payment method is saved. This is likely unrelated to COT.~ This just happens while buying subscriptions – payment methods will be always saved in that case.
- In "Shopper - Subscriptions - Change payment method to new card" and related tests, the payment method is not updated after landing on Subscription view, even though alert says so. It's likely unrelated to COT.

**To do:**

- [ ] Confirm and create GH issue for the changing payment method issue.
- [ ] Set Stripe compatible with HPOS.

**Comment by reykjalin** [comment]
I switched my local dev store to HPOS and subscriptions don't seem to work for me.

**Store setup**

1. WordPress v6.1.1
2. WooCommerce v7.3.0
3. WooCommerce Stripe `develop` (https://github.com/woocommerce/woocommerce-gateway-stripe/commit/d9ed8a9be5f67c6c4290032fea30b67d6617b46f) with the following patch applied:

```diff
diff --git a/woocommerce-gateway-stripe.php b/woocommerce-gateway-stripe.php
index 6a0a76c8..8b2a9359 100644
--- a/woocommerce-gateway-stripe.php
+++ b/woocommerce-gateway-stripe.php
@@ -777,7 +777,7 @@ add_action(
        'before_woocommerce_init',
        function() {
                if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
-                       \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, false );
+                       \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
                }
        }
 );
```

4. WooCommerce Blocks v9.3.0 (Although I used the shortcode checkout)
5. WooCommerce Subscriptions v4.7.0 with the following patch applied:

```diff
--- a/woocommerce-subscriptions.php
+++ b/woocommerce-subscriptions.php
add_action(
    'before_woocommerce_init',
    function() {
        if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
-           \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, false );
+           \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
        }
    }
);
```
6. **UPE is enabled in Stripe**.

**Testing flow**

1. Add a single (simple) subscription product with no free trial period to the cart.
2. Go to the shortcode checkout.
3. Complete checkout with the non-3DS test card `4242 4242 4242 4242`.
4. Go to **wp-admin** > **WooCommerce** > **Orders** and open the order details for the order.
5. Order contains a subscription and was successfully processed. **Overall looks normal**.
6. [Stripe dashboard](https://dashboard.stripe.com) shows the payment was collected successfully.
7. Go to **wp-admin** > **WooCommerce** > **Subscriptions** and notice **the subscription is missing**.

| Order details | Subscriptions list |
|:---:|:---:|
| <img width="1147" alt="image" src="https://user-images.githubusercontent.com/13835680/212786941-835fb7f2-3cd7-41c2-b8ea-ce0e56fc4f94.png"> | <img width="891" alt="image" src="https://user-images.githubusercontent.com/13835680/212786988-0025c5ca-a4b4-4231-82eb-636d5041835b.png"> |

8. Go to **wp-admin** > **WooCommerce** > **Orders** and open the order details for the order. From there open the details for the subscription order. Everything looks normal.
9. Go to **wp-admin** > **Tools** > **Scheduled Actions**. Search for the renewal action and run it.
10. **Renewal action for the order fails**.

<details><summary>Logs</summary>

```
2023-01-17T00:57:17+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
Info: Begin processing subscription payment for order 191 for the amount of 10.00
====End Log====


2023-01-17T00:57:17+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
payment_intents request: Array
(
    [amount] => 1000
    [currency] => usd
    [description] => Stripe - Order 191
    [metadata] => Array
        (
            [customer_name] => Tester Testerson
            [customer_email] => any@example.com
            [order_id] => 191
            [site_url] => http://stripe.test
            [payment_type] => recurring
        )

    [payment_method_types] => Array
        (
            [0] => card
            [1] => link
        )

    [off_session] => true
    [confirm] => true
    [confirmation_method] => automatic
    [statement_descriptor] => THRLKSN
    [customer] => cus_N7lr0MremTl2lc
    [shipping] => Array
        (
            [name] => Tester Testerson
            [address] => Array
                (
                    [line1] => 60 29th St
                    [line2] => #343
                    [city] => San Francisco
                    [country] => US
                    [postal_code] => 94110
                    [state] => CA
                )

        )

)

====End Log====


2023-01-17T00:57:19+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
Info: Begin processing subscription payment for order 191 for the amount of 10.00
====End Log====


2023-01-17T00:57:19+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
payment_intents request: Array
(
    [amount] => 1000
    [currency] => usd
    [description] => Stripe - Order 191
    [metadata] => Array
        (
            [customer_name] => Tester Testerson
            [customer_email] => any@example.com
            [order_id] => 191
            [site_url] => http://stripe.test
            [payment_type] => recurring
        )

    [payment_method_types] => Array
        (
            [0] => card
            [1] => link
        )

    [off_session] => true
    [confirm] => true
    [confirmation_method] => automatic
    [statement_descriptor] => THRLKSN
    [customer] => cus_N7lr0MremTl2lc
    [shipping] => Array
        (
            [name] => Tester Testerson
            [address] => Array
                (
                    [line1] => 60 29th St
                    [line2] => #343
                    [city] => San Francisco
                    [country] => US
                    [postal_code] => 94110
                    [state] => CA
                )

        )

)

====End Log====


2023-01-17T00:57:21+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
Info: Begin processing subscription payment for order 191 for the amount of 10.00
====End Log====


2023-01-17T00:57:21+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
payment_intents request: Array
(
    [amount] => 1000
    [currency] => usd
    [description] => Stripe - Order 191
    [metadata] => Array
        (
            [customer_name] => Tester Testerson
            [customer_email] => any@example.com
            [order_id] => 191
            [site_url] => http://stripe.test
            [payment_type] => recurring
        )

    [payment_method_types] => Array
        (
            [0] => card
            [1] => link
        )

    [off_session] => true
    [confirm] => true
    [confirmation_method] => automatic
    [statement_descriptor] => THRLKSN
    [customer] => cus_N7lr0MremTl2lc
    [shipping] => Array
        (
            [name] => Tester Testerson
            [address] => Array
                (
                    [line1] => 60 29th St
                    [line2] => #343
                    [city] => San Francisco
                    [country] => US
                    [postal_code] => 94110
                    [state] => CA
                )

        )

)

====End Log====


2023-01-17T00:57:25+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
Info: Begin processing subscription payment for order 191 for the amount of 10.00
====End Log====


2023-01-17T00:57:25+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
payment_intents request: Array
(
    [amount] => 1000
    [currency] => usd
    [description] => Stripe - Order 191
    [metadata] => Array
        (
            [customer_name] => Tester Testerson
            [customer_email] => any@example.com
            [order_id] => 191
            [site_url] => http://stripe.test
            [payment_type] => recurring
        )

    [payment_method_types] => Array
        (
            [0] => card
            [1] => link
        )

    [off_session] => true
    [confirm] => true
    [confirmation_method] => automatic
    [statement_descriptor] => THRLKSN
    [customer] => cus_N7lr0MremTl2lc
    [shipping] => Array
        (
            [name] => Tester Testerson
            [address] => Array
                (
                    [line1] => 60 29th St
                    [line2] => #343
                    [city] => San Francisco
                    [country] => US
                    [postal_code] => 94110
                    [state] => CA
                )

        )

)

====End Log====


2023-01-17T00:57:30+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
Info: Begin processing subscription payment for order 191 for the amount of 10.00
====End Log====


2023-01-17T00:57:30+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
payment_intents request: Array
(
    [amount] => 1000
    [currency] => usd
    [description] => Stripe - Order 191
    [metadata] => Array
        (
            [customer_name] => Tester Testerson
            [customer_email] => any@example.com
            [order_id] => 191
            [site_url] => http://stripe.test
            [payment_type] => recurring
        )

    [payment_method_types] => Array
        (
            [0] => card
            [1] => link
        )

    [off_session] => true
    [confirm] => true
    [confirmation_method] => automatic
    [statement_descriptor] => THRLKSN
    [customer] => cus_N7lr0MremTl2lc
    [shipping] => Array
        (
            [name] => Tester Testerson
            [address] => Array
                (
                    [line1] => 60 29th St
                    [line2] => #343
                    [city] => San Francisco
                    [country] => US
                    [postal_code] => 94110
                    [state] => CA
                )

        )

)

====End Log====


2023-01-17T00:57:30+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
Info: Begin processing subscription payment for order 191 for the amount of 10.00
====End Log====


2023-01-17T00:57:30+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
payment_intents request: Array
(
    [amount] => 1000
    [currency] => usd
    [description] => Stripe - Order 191
    [metadata] => Array
        (
            [customer_name] => Tester Testerson
            [customer_email] => any@example.com
            [order_id] => 191
            [site_url] => http://stripe.test
            [payment_type] => recurring
        )

    [payment_method_types] => Array
        (
            [0] => card
            [1] => link
        )

    [off_session] => true
    [confirm] => true
    [confirmation_method] => automatic
    [statement_descriptor] => THRLKSN
    [customer] => cus_N7lr0MremTl2lc
    [shipping] => Array
        (
            [name] => Tester Testerson
            [address] => Array
                (
                    [line1] => 60 29th St
                    [line2] => #343
                    [city] => San Francisco
                    [country] => US
                    [postal_code] => 94110
                    [state] => CA
                )

        )

)

====End Log====


2023-01-17T00:57:31+00:00 DEBUG 
====Stripe Version: 7.0.2====
====Start Log====
Error: stdClass Object
(
    [error] => stdClass Object
        (
            [code] => payment_intent_unexpected_state
            [doc_url] => https://stripe.com/docs/error-codes/payment-intent-unexpected-state
            [message] => You cannot confirm this PaymentIntent because it's missing a payment method. You can either update the PaymentIntent with a payment method and then confirm it again, or confirm it again directly with a payment method.
            [request_log_url] => https://dashboard.stripe.com/test/logs/req_eDMB8MdnINP1i4?t=1673917055
            [type] => invalid_request_error
        )

)

====End Log====
```
</details>

The most relevant bit from the logs is this:

```
Error: stdClass Object
(
    [error] => stdClass Object
        (
            [code] => payment_intent_unexpected_state
            [doc_url] => https://stripe.com/docs/error-codes/payment-intent-unexpected-state
            [message] => You cannot confirm this PaymentIntent because it's missing a payment method. You can either update the PaymentIntent with a payment method and then confirm it again, or confirm it again directly with a payment method.
            [request_log_url] => https://dashboard.stripe.com/test/logs/req_eDMB8MdnINP1i4?t=1673917055
            [type] => invalid_request_error
        )

)
```

So it looks like there is no payment method assigned to the renewal payments? Not sure exactly what's going on.

11. The new order in **wp-admin** > **WooCommerce** > **Orders** is `Pending payment`.
12. There is now a subscription that's set to `On Hold` in **wp-admin** > **WooCommerce** > **Subscriptions**.
13. [Stripe dashboard](https://dashboard.stripe.com) shows no attempt at creating any sort of payment intent in the Payments List.
14. The logs in the Stripe dashboard show 6 failed attempts:

<img width="651" alt="image" src="https://user-images.githubusercontent.com/13835680/212787548-ec69c017-59db-44cb-9007-2e77678c89ac.png">

**Comment by reykjalin** [comment]
Same thing happens on a fresh store, except now renewals work but there's no Subscription added to **wp-admin** > **WooCommerce** > **Subscription** 🤔

**Comment by reykjalin** [comment]
Hmmm, after I deleted the payment methods that were initially saved during the first few subscription trials to reduce the list of payment methods in the database the renewals started failing again, likely because no payment method was saved:

<img width="1185" alt="image" src="https://user-images.githubusercontent.com/13835680/212819171-d98b97f9-8260-40c4-8f70-d40900a45da7.png">

**This is with UPE enabled**. It seems like UPE + HPOS doesn't save payment methods? Even when I checkout with a simple product and tick "save payment method" there is still no payment method added.

Adding a payment method through the payment method screen caused them all the previous trial attempts to appear:

<img width="1185" alt="image" src="https://user-images.githubusercontent.com/13835680/212819466-9a70aad7-c8a0-4410-a2c1-c08a8ad0328e.png">

**Comment by reykjalin** [comment]
Switching back to the non-UPE checkout has renewals working again, but there are still no Subscriptions showing up in the subscriptions list.

**Comment by reykjalin** [comment]
Using UPE with Blocks checkout works the same as non-UPE with the blocks checkout, i.e. payments and renewals successful, but no Subscriptions show up in the subscriptions list.

So in my tests I think this is the array of support for subs, at least as long as the Subs plugin wraps up the support by making sure the Subscriptions show up in the Subscriptions list:

| | Shortcode | Blocks |
|:---:|:---:|:---:|
| **Non-UPE Checkout** | ✅ | ✅ |
| **UPE Checkout** | ❌ | ✅ |

The UPE Checkout with the shortcode block is likely failing because the payment method is not being saved for some reason, which I think is a problem I've seen before sometime 🤔

**Comment by reykjalin** [comment]
When I use the latest subs testing release from pdjTHR-1Fu-p2#comment-2572 everything seems to work actually, so the issue is only present with v4.7.0 of subs (which we already knew to be incompatible).

I'm just glad the shortcode+UPE issue is an issue in subs, not Stripe 😌

**Comment by cesarcosta99** [comment]
@reykjalin In a fresh store using the very same setup you used, except Stripe, where I used 7.0.2, I could reproduce partially what you described [here](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2407#issuecomment-1384712429). I got the same until step 9 – the action runs fine and the order is processed and paid correctly, although, no subscriptions show up still.

After using the test release you mentioned it worked well 🙌  Replacing the subs 4.7.0 plugin with the test version in the same store step fixes the subscription list as well.

**Comment by reykjalin** [comment]
Doing another battery of tests with “Use the WooCommerce orders tables” option enabled:

## Store setup

1. WordPress 6.1.1
2. WooCommerce 7.3.0
4. WooCommerce Stripe `develop` (1c36361ad485c962f5c7d27578a1ae8f8d2c9356)
5. WooCommerce Subscriptions Test Release from pdjTHR-1Fu-p2#comment-2572

## Flows

> **Note**
> All tests done with HPOS **enabled** and  “Keep the posts table and the orders tables synchronized" **disabled**.

> **Note**
> Testing was done on a fresh store, but before starting the flows I processed 1 payment with HPOS turned off in an attempt to simulate a store that's migrating from WP tables to HPOS.

- [x] Merchant - Onboarding - Manual plugin installation and setup
- [x] Shopper - Checkout - Checkout with normal credit card
- [x] Shopper - Checkout - Save card at checkout
    - UPE: ❌ This did not work; the card wasn't saved.
    - No UPE: ❌ This did not work; the card wasn't saved.
        - Although the cards do show up when you go back to the Checkout page. If you go to My Account > Payment Methods they do not show up (unless you've already gone back to the Checkout page).
        - This is likely because we have some code to fetch and update payment methods from Stripe when loading the Checkout page.
- [x] Shopper - Checkout - Checkout with saved card
- [x] Shopper - Checkout - Checkout with SCA card
- [x] Shopper - Checkout - Checkout failures (with various cards)
- [x] Shopper - Checkout - Checkout with Payment Request
- [ ] Shopper - Checkout - Checkout with Payment Request (3DS card)
  - Can’t test this one.
- [x] Shopper - Payment methods - Add new payment method
- [x] Shopper - Payment methods - Delete payment method
- [x] Merchant - Order - Manual capture
- [x] Merchant - Order - Full refund
- [x] Merchant - Order - Partial refund
- [x] Merchant - Order - Refund failure
- [x] Shopper - Subscriptions - Purchase subscription product
- [x] Shopper - Subscriptions - Purchase free trial subscription
- [x] Shopper - Subscriptions - Purchase multiple subscriptions (same schedule)
- [x] Shopper - Subscriptions - Purchase multiple subscriptions (differing schedules)
- [x] Shopper - Subscriptions - Renew subscription
- [x]  Shopper - Subscriptions - Change payment method to new card
    - UPE: ❌ Success message shows up, but the "Payment" box is empty.
    - No UPE: ❌ Success message shows up, but the "Payment" box is empty.
- [x] Shopper - Subscriptions - Change payment method to saved card
    - UPE: ❌ Success message shows up, but the "Payment" box is empty.
    - No UPE: ❌ Success message shows up, but the "Payment" box is empty.
- [x] Shopper - Subscriptions - Set default payment method
    - UPE: ❌ Success message shows up, but the "Payment" box is empty.
    - No UPE: ❌ Success message shows up, but the "Payment" box is empty.
- [x] Shopper - Subscriptions - Change default payment method
    - UPE: ❌ unable to complete steps from previous step, so unable to perform this test.
    - No UPE: ❌ unable to complete steps from previous step, so unable to perform this test.
- [x] Merchant - Subscriptions - Renew subscription automatically
    - UPE: ❌ Failed to save card during checkout, so the renewal payment failed.
    - No UPE: ❌ Failed to save card during checkout, so the renewal payment failed.
- [x] Merchant - Subscriptions - Renew subscription manually
- [x] Merchant - WooCommerce Blocks - Create a Cart page with the Cart Block
- [x] Merchant - WooCommerce Blocks - Create a Checkout page with the Checkout Block
- [x] Shopper - WooCommerce Blocks - Checkout with normal credit card
- [x] Shopper - WooCommerce Blocks - Checkout with SCA card
- [x] Shopper - WooCommerce Blocks - Checkout failures (with various cards)
- [x] Shopper - WooCommerce Blocks - Checkout with Payment Request
- [x] Shopper - WooCommerce Blocks - Save card at checkout
    - UPE: ❌ This did not work; the card wasn't saved.
    - No UPE: ✅ Works.
- [x] Shopper - WooCommerce Blocks - Checkout with saved card
- [x] Shopper - WooCommerce Blocks - Purchase subscription product
- [x] Shopper - WooCommerce Blocks - Purchase free trial subscription
- [x] Shopper - WooCommerce Blocks - Purchase multiple subscriptions (same schedule)
- [x] Shopper - WooCommerce Blocks - Purchase multiple subscriptions (differing schedules)
- [ ] Shopper - APMs - Checkout with iDEAL
- [ ] Shopper - APMs - Checkout with Sofort
- [ ] Shopper - APMs - Checkout with giropay
- [ ] Shopper - APMs - Checkout with EPS
- [ ] Shopper - APMs - Checkout with Bancontact
- [ ] Shopper - APMs - Checkout with Przelewy24
- [ ] Shopper - APMs - Checkout with SEPA
- [ ] Shopper - APMs - Checkout with Multibanco
- [ ] Shopper - APMs - Checkout with Alipay
- [ ] Shopper - APMs - Checkout with Boleto
- [ ] Shopper - APMs - Checkout with OXXO

## Result

There seems to be an issue with saving cards when UPE is enabled causing cascading issues for subscriptions. The first payment seems to work, subscriptions are created correctly, and renewal jobs are properly instantiated. The only thing missing is payment method saving functionality.

## Bugs found

1. Saving cards is flaky **at best**. For some reason it was working for me last night but is not today. I'm not sure why.
    - This consequently makes it so subscriptions don't work. Not sure if the fault is with the plugin, subscriptions, HPOS, or any combination of all three.


## To do

- [ ] Open issue to fix the issue with saving payment methods.

**Comment by cesarcosta99** [comment]
@reykjalin I retested all of the items that failed for you. The following **worked** for me under the same store setup (with Stripe 7.0.2):

- Shopper - Checkout - Save card at checkout
  - UPE: ❌ This did not work; the card wasn't saved.
  - No UPE: ❌ This did not work; the card wasn't saved.
- Shopper - WooCommerce Blocks - Save card at checkout
  - UPE: ❌ This did not work; the card wasn't saved.

Can you double-check the ones above in a fresh store to confirm?

I spun up a new store with Stripe 7.0.2 and HPOS disabled and all tests under "Shopper - Subscriptions" that failed with HPOS enabled passed in this store. I'm investigating this.

**Comment by cesarcosta99** [comment]
Created #2531 to track failing critical flows.

**Comment by asumaran** [comment]
> @reykjalin I retested all of the items that failed for you. The following **worked** for me under the same store setup (with Stripe 7.0.2):
> 
> * Shopper - Checkout - Save card at checkout
>   
>   * UPE: ❌ This did not work; the card wasn't saved.
>   * No UPE: ❌ This did not work; the card wasn't saved.
> * Shopper - WooCommerce Blocks - Save card at checkout
>   
>   * UPE: ❌ This did not work; the card wasn't saved.


@cesarcosta99 @reykjalin This worked for me as well. With HPOS enabled and UPE enabled/disabled and using shortcode based checkout or checkout block.
